### ISSUE TITLE
`browser.element()` does not work after a `browser.waitUntil` is invoked

### ISSUE THREAD

**Issue by khirakawa** [webdriverio/webdriverio#1690]
## The problem

`browser.element()` does not work after a `browser.waitUntil` is invoked

## Environment

* WebdriverIO version: 4.3.0
* Node.js version: 4.3.0
* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html): wdio testrunner
* if wdio testrunner, running synchronous or asynchronous tests: synchronous
* Additional wdio packages used (if applicable): none

## Details

```
describe('test', function () {
    it('should do something', function () {
        browser.url('https://google.com');

        // This returns with an element
        console.log(browser.element('[data-jiis="cc"]'));

        browser.waitUntil(function () {
            try {
                const element = browser.element('[data-jiis="cc"]')
                return !!element.value;
            } catch (e) {
                return false;
            }
        });

        // This does not
        console.log(browser.element('[data-jiis="cc"]'));
    });
});
```

In the code above, the first console.log works and returns with an element.  However, the second console.log will return with `NoSuchElement`.  The only difference between the two, is that there's a `waitUntil` in between that waits for the same element to exist.

## Link to Selenium/WebdriverIO logs

https://gist.github.com/khirakawa/86123dc7284e40696861e972cd8141c8

## Code To Reproduce Issue [ Good To Have ]

See above

**Comment by tiffon** [comment]
I've also run into this issue. Looks like the final `browser.element(...)` is scoped to the previous call's results.

An excerpt of the logs:

```
        [01:31:23]  COMMAND POST     "/wd/hub/session/7ce5a828-5f43-45ab-a210-1791b8ec37bb/element"
        [01:31:23]  DATA        {"using":"css selector","value":"[data-jiis=\"cc\"]"}
        [01:31:23]  RESULT      {"ELEMENT":"0"}
        { state: 'success',
          sessionId: '7ce5a828-5f43-45ab-a210-1791b8ec37bb',
          hCode: 1896067273,
          value: { ELEMENT: '0' },
          class: 'org.openqa.selenium.remote.Response',
          selector: '[data-jiis="cc"]',
          _status: 0 }
        [01:31:23]  COMMAND POST     "/wd/hub/session/7ce5a828-5f43-45ab-a210-1791b8ec37bb/element"
        [01:31:23]  DATA        {"using":"css selector","value":"[data-jiis=\"cc\"]"}
        [01:31:23]  RESULT      {"ELEMENT":"0"}
--->    [01:31:23]  COMMAND POST     "/wd/hub/session/7ce5a828-5f43-45ab-a210-1791b8ec37bb/element/0/element"
        [01:31:23]  DATA        {"using":"css selector","value":"[data-jiis=\"cc\"]"}
        { type: 'NoSuchElement',
          message: 'An element could not be located on the page using the given search parameters.',
          state: 'failure',
          sessionId: '7ce5a828-5f43-45ab-a210-1791b8ec37bb',
          value: null,
          selector: '[data-jiis="cc"]',
          _status: 7 }
```

**Comment by tiffon** [comment]
FYI, you can work around this by scoping the `element` call in the `waitUntil(...)` to another element:

```javascript
describe('test', function () {
    it('should do something', function () {
        browser.url('https://google.com');

        // This returns with an element
        console.log(browser.element('[data-jiis="cc"]'));
        const body = browser.element('body');
        browser.waitUntil(function () {
            try {
                return !!body.element('[data-jiis="cc"]').value;
            } catch (e) {
                return false;
            }
        });

        // This does not
        console.log(browser.element('[data-jiis="cc"]'));
    });
});
```

Log excerpt:
```

[01:43:55]  COMMAND	POST 	 "/wd/hub/session/e8dc0b35-33f7-482f-94e8-22c7d45f1046/element"
[01:43:55]  DATA		{"using":"css selector","value":"[data-jiis=\"cc\"]"}
[01:43:55]  RESULT		{"ELEMENT":"0"}
{ state: 'success',
  sessionId: 'e8dc0b35-33f7-482f-94e8-22c7d45f1046',
  hCode: 2072065264,
  value: { ELEMENT: '0' },
  class: 'org.openqa.selenium.remote.Response',
  selector: '[data-jiis="cc"]',
  _status: 0 }
[01:43:55]  COMMAND	POST 	 "/wd/hub/session/e8dc0b35-33f7-482f-94e8-22c7d45f1046/element"
[01:43:55]  DATA		{"using":"css selector","value":"body"}
[01:43:55]  RESULT		{"ELEMENT":"1"}
[01:43:55]  COMMAND	POST 	 "/wd/hub/session/e8dc0b35-33f7-482f-94e8-22c7d45f1046/element/1/element"
[01:43:55]  DATA		{"using":"css selector","value":"[data-jiis=\"cc\"]"}
[01:43:55]  RESULT		{"ELEMENT":"2"}
[01:43:55]  COMMAND	POST 	 "/wd/hub/session/e8dc0b35-33f7-482f-94e8-22c7d45f1046/element"
[01:43:55]  DATA		{"using":"css selector","value":"[data-jiis=\"cc\"]"}
[01:43:55]  RESULT		{"ELEMENT":"0"}
{ state: 'success',
  sessionId: 'e8dc0b35-33f7-482f-94e8-22c7d45f1046',
  hCode: 1556940643,
  value: { ELEMENT: '0' },
  class: 'org.openqa.selenium.remote.Response',
  selector: '[data-jiis="cc"]',
  _status: 0 }
```

**Comment by tiffon** [comment]
This actually does work around the issue:

```javascript
function waitUntil(fn, timeout, invert) {
  const expect = !invert;
  const expire = Date.now() + (timeout > 0 ? timeout : 10000);
  let ok = fn() === expect;
  while (Date.now() < expire && !ok) {
    browser.pause(200);
    ok = fn() === expect;
  }
  if (!ok) {
    throw new Error('doWaitUntil failed');
  }
}
```

But, if it's added as a command it has the same problem as `browser.waitUntil(...)`.

**Comment by khirakawa** [comment]
@tiffon Thanks for the workarounds!  I'm curious to know why [your first workaround](https://github.com/webdriverio/webdriverio/issues/1690#issuecomment-258662102) doesn't actually solve the problem.  I tried it out and it seemed to work...

> But, if it's added as a command it has the same problem as browser.waitUntil(...).

I'm not quite familiar with how `browser.waitUntil` works.  Could you explain why this is the case?

**Comment by christian-bromann** [comment]
@khirakawa @tiffon this should be fixed by #1702. Please update the framework adapter you are using. If the error still occurs, let me know and I will reopen.
